src/bootman/bootman.c

```diff
@@ -446,6 +458,174 @@ bool boot_manager_set_uname(BootManager *self, const char *uname)
         return self->have_sys_kernel;
 }
 
+static int initrd_nodep_name_cmp(const void *file1, const void *file2)
+{
+        return strcmp((char*)file1, (char*)file2);
+}
+
+void boot_manager_get_initrd_nodep(BootManager *self)
+{
+        DIR *dir = NULL;
+        struct dirent *ent = NULL;
+        struct stat st = { 0 };
+        NcArray *initrd_nodep = NULL;
+        if (!self || !self->initrd_nodep_dir) {
+                return;
+        }
+
+        initrd_nodep = nc_array_new();
+        OOM_CHECK_RET(initrd_nodep,);
+
+        dir = opendir(self->initrd_nodep_dir);
+        if (!dir) {
+                LOG_ERROR("Error opening %s: %s", self->initrd_nodep_dir, strerror(errno));
+                nc_array_free(&initrd_nodep, NULL);
+                return;
+        }
+
+        while ((ent = readdir(dir)) != NULL) {
+        char *initrd_name = NULL;
+                autofree(char) *path = NULL;
+
+                path = string_printf("%s/%s", self->initrd_nodep_dir, ent->d_name);
+
+                if (strstr(ent->d_name, "img.gz\0") == NULL) {
+                        continue;
+                }
+
+                /* Some kind of broken link */
+                if (lstat(path, &st) != 0) {
+                        continue;
+                }
+
+                /* Regular only */
+                if (!S_ISREG(st.st_mode)) {
+                        continue;
+                }
+
+                /* empty files are skipped too */
+                if (st.st_size == 0) {
+                        continue;
+                }
+
+                initrd_name = strdup(ent->d_name);
+                if (!nc_array_add(initrd_nodep, initrd_name)) {
+                        DECLARE_OOM();
+                        abort();
+                }
+        }
+        if (initrd_nodep->len > 0) {
+                nc_array_qsort(initrd_nodep, initrd_nodep_name_cmp);
+                self->initrd_nodep = strdup(initrd_nodep->data[0]);
+        }
+        nc_array_free(&initrd_nodep, free);
+        closedir(dir);
+}
+
+void boot_manager_copy_initrd_nodep(BootManager *self)
```

Comment:
This should not be void, if it fails it needs to fail properly and pass the failure back up